### ISSUE TITLE
Cargo check does not report `error: symbol `foo` is already defined`

### ISSUE THREAD

**Issue by mcarton** [rust-lang/rust#61330]
**Problem**
Some errors are not reported by `cargo check`.

**Steps**
For example, here `cargo check` succeeds:
```rust
% cargo check
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
% cat src/main.rs 
#[no_mangle]
fn test() {}

mod foo {
    #[no_mangle]
    fn test() {}
}

fn main() {
    println!("Hello, world!");
}
```

But `cargo build` does fail:
```rust
% cargo build
   Compiling asdf v0.1.0 (/tmp/asdf)
error: symbol `test` is already defined
 --> src/main.rs:6:5
  |
6 |     fn test() {}
  |     ^^^^^^^^^^^^

error: aborting due to previous error

error: Could not compile `asdf`.

To learn more, run the command again with --verbose.
```

**Notes**

Output of `cargo version`:
```
% cargo --version      
cargo 1.34.0 (6789d8a0a 2019-04-01)
% rustc --version
rustc 1.34.2 (6c2484dc3 2019-05-13)
% rustup --version
rustup 1.18.1 (2019-04-26)
```

**Comment by ehuss** [comment]
Moved to rust-lang/rust.

I think this is to be expected. `cargo check` performs all steps up to translation.  Some errors are only detected during later phases.  I don't know if it is feasible to check errors like this.

**Comment by mcarton** [comment]
I understand why it fails, but the intent of `cargo check` seems to be that if `cargo check` succeeds, then `cargo build` will as well, however this is not true, and doesn't seem to be explicitly documented anywhere.

```
$ cargo check --help
cargo-check 
Check a local package and all of its dependencies for errors
```

**Comment by Mark-Simulacrum** [comment]
I think this is essentially unavoidable -- there's a reason cargo build and cargo check are separate; one can do strictly more error checking than the other. I'm going to close this as not a bug.
